org.opent2t.sample.hub.superpopular/com.smartthings.hub/js/groovy/SmartApp.groovy

```diff
@@ -129,290 +146,239 @@ def registerChangeHandler(myList) {
+						log.info "Update: ${device.displayName}, [${command}, ${value}]"
+						device.setHeatingSetpoint(value)
+						break
+					default:
+						httpError(501, "this mode: ${device.currentThermostatMode} does not allow changing thermostat setpoint.")
+						break
+				}
+			}else if (!device) {
+				log.error "updateDevice, Device not found"
+				httpError(404, "Device not found")
+			} else if (!device.hasCommand(command)) {
+				log.error "updateDevice, Device does not have the command"			
+				httpError(404, "Device does not have such command")
+			} else {
+				if (command == "setColor") {
+					log.info "Update: ${device.displayName}, [${command}, ${value}]"
+					device."$command"(hex: value)
+				} else if(value.isNumber()) {
+					def intValue = value as Integer
+					log.info "Update: ${device.displayName}, [${command}, ${intValue}(int)]"
+					device."$command"(intValue)
+				} else if (value){
+					log.info "Update: ${device.displayName}, [${command}, ${value}]"
+					device."$command"(value)
+				} else {
+					log.info "Update: ${device.displayName}, [${command}]"
+					device."$command"()
+				}
+			}
+		}
 	}
 }
 
 /*** Private Functions ***/
 
 //Return current location mode info
 private getLocationModeInfo() {
-    return [mode: location.mode, supported: location.modes.name]
+	return [mode: location.mode, supported: location.modes.name]
 }
 
 //Map each device to a type given it's capabilities
 private getDeviceType(device) {
 	def deviceType
-    def caps = device.capabilities
-    log.debug "capabilities: [${device}, ${caps}]"
-    log.debug "supported commands: [${device}, ${device.supportedCommands}]"
+	def caps = device.capabilities
+	log.debug "capabilities: [${device}, ${caps}]"
+	log.debug "supported commands: [${device}, ${device.supportedCommands}]"
 	caps.each {
-        switch(it.name.toLowerCase())
-        {
-        	case "switch":
+		switch(it.name.toLowerCase())
+		{
+			case "switch":
             	deviceType = "switch"
-            	break
-            case "switch level":
-            	deviceType = "light"
-            	break
-            case "contact sensor":
-            	deviceType = "contactSensor"
-            	break
-            case "garageDoorControl":
-            	deviceType = "garageDoor"
-            	break
-            case "lock":
-            	deviceType = "lock"
-            	break
-            case "video camera":
-            	deviceType = "camera"
-            	break
-            case "motion sensor":
-            	deviceType = "motionSensor"
-            	break
-            case "presence sensor":
-            	deviceType = "presenceSensor"
-            	break
-            case "thermostat":
-            	deviceType = "thermostat"
-            	break
-            case "water sensor":
-            	deviceType = "waterSensor"
-            	break   
-            default:
-            	break
-        }
-    }
-    return deviceType
+				if (caps.any{it.name.toLowerCase() == "power meter"}){
```

Comment:
+1 -- unless in groovy, the 'it' implicit parameter gets re-scoped to the current value in the new loop?? either ways this is too confusing.. both logic (switch capablity doesnt mean switch device type; it could also mean 'light' depending on presence of another cap (switch level) in the returned cap list) as well as confusing as far as implementation (this use of default 'it' parameter is confusing) 

---
In reply to: [100167232](https://github.com/openT2T/translators/pull/152#discussion_r100167232) [](ancestors = 100167232)